Allow local users with an expiry date to be created

rpluem-vf opened the pull request:
The luseradd / lusermod commands do not support the -e option. Set
the expiry time in this case via lchage after the user was
created / modified.

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes: #71942

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
user

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

rpluem-vf committed:
Allow local users with an expiry date to be created

The luseradd / lusermod commands do not support the -e option. Set
the expiry time in this case via lchage after the user was
created / modified.

Fixes: #71942

rpluem-vf committed:
Add missing 'self'

rpluem-vf committed:
Remove home directory as well when removing the user

rpluem-vf committed:
Use local_ansibulluser instead of ansibulluser username for local tests

rpluem-vf committed:
In case of failure remove the user

rpluem-vf committed:
Indent correctly

rpluem-vf committed:
Revert "Indent correctly"

This reverts commit b3bf79871b4f05b3fabfd23bb60e1b93b9eb4c5e.

rpluem-vf committed:
Revert "In case of failure remove the user"

This reverts commit e9f4c63de787e25cf249bf16c1ffa8e7e4b61138.

rpluem-vf committed:
Convert math.floor to an int

In Python3 math.floor returns an integer whereas Python2 returns a float.
Hence always convert the result of math.floor to an int to ensure that
lexpires is an integer.

rpluem-vf commented:
ready_for_review

samdoran commented on changelogs/fragments/fix_ansible_issue_71942.yaml:
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - >
+     user - Local users with an expiry date cannot be created as the luseradd /
+     lusermod commands do not support the -e option. Set the expiry time in
+     this case via lchage after the user was created / modified. (https://github.com/ansible/ansible/issues/71942)
```
```suggestion
     user - Local users with an expiry date cannot be created as the ``luseradd`` /
     ``lusermod`` commands do not support the ``-e`` option. Set the expiry time in
     this case via ``lchage`` after the user was created / modified. (https://github.com/ansible/ansible/issues/71942)
```

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -674,7 +676,22 @@ def create_user_useradd(self):
+                lexpires = -1
+            else:
+                # Convert seconds since Epoch to days since Epoch
+                lexpires = int(math.floor(float(self.module.params['expires']))) // 86400
```
I'm not sure the `float` conversion is necessary. The `expires` parameter is specified as `type='float'` which means it has already gone through `check_type_float()` which means it has already been converted to a `float`.

I had to look into why `int(math.floor())` was necessary. `math.floor()` returns a `float` on Python 2 and an `int` in Python 3. I learned a bit from this line, so thanks for that!

rpluem-vf replied:
I learned as well :-). I stumbled across this as the tests failed on some platforms before https://github.com/ansible/ansible/pull/72022/commits/1695b5ed8b33afee8fba34ae759eaaeb98abc15b and the difference was as you state that the failing platforms ran the tests on Python 2 while the succeeding ones ran on Python 3. Hence the great wealth of platforms tested was very helpful in detecting this issue.
Done in afd6e08.

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -674,7 +676,22 @@ def create_user_useradd(self):
+            else:
+                # Convert seconds since Epoch to days since Epoch
+                lexpires = int(math.floor(float(self.module.params['expires']))) // 86400
+            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', str(lexpires), self.name])
```
```suggestion
            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', to_native(lexpires), self.name])
```

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -801,16 +820,23 @@ def modify_user_usermod(self):
-                    cmd.append(time.strftime(self.DATE_FORMAT, self.expires))
+                    if self.local:
+                        # Convert seconds since Epoch to days since Epoch
+                        lexpires = int(math.floor(float(self.module.params['expires']))) // 86400
```
Same here: I don't believe the inner `float()` conversion is necessary.

rpluem-vf replied:
I guess you are correct. Will change it and let us see what the testing says.

rpluem-vf replied:
Done in https://github.com/ansible/ansible/pull/72022/commits/afd6e08b4fa3ef2e002f67dfc7e0efb6fbcc15f2

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -830,7 +856,17 @@ def modify_user_usermod(self):
+            return (rc, err, out)
+
+        if lexpires is not None:
+            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', str(lexpires), self.name])
```
```suggestion
            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', to_native(lexpires), self.name])
```

samdoran commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -1135,3 +1135,336 @@
   when: ansible_facts.system in ['Linux']
   tags:
     - user_test_local_mode
+
```
These tests are great. This `main.yml` file is a rather huge and I plan to break it up at some point. Could you move these tests into a separate tasks file and use `import_tasks` to make my future work a bit easier?

rpluem-vf replied:
Done in https://github.com/ansible/ansible/pull/72022/commits/d5c473e34986277a0cd35390608e4cdca2a1eb5f

samdoran reviewed (changes requested):
The change is good overall. Just a few minor changes are needed.

rpluem-vf committed:
Mark options and commands as such

Co-authored-by: Sam Doran <sdoran@redhat.com>

rpluem-vf committed:
Use to_native instead of str for correct string conversion

Co-authored-by: Sam Doran <sdoran@redhat.com>

rpluem-vf committed:
Use to_native instead of str for correct string conversion

Co-authored-by: Sam Doran <sdoran@redhat.com>

rpluem-vf committed:
Remove unneeded float conversion

rpluem-vf committed:
Move local expires tests in a separate file

Move local expires tests in a separate file and import the tasks to the
main.yml to keep main.yml smaller.

rpluem-vf committed:
Fix typo

rpluem-vf reviewed (commented):
Hopefully things are fine now.

samdoran reviewed (commented):
Thank you for addressing my comments so quickly.

samdoran commented:
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs) against `stable-2.9` and `stable-2.10`.

rpluem-vf commented:
stable-2.10:  #72085
stable-2.9: #72086
